### ISSUE TITLE
segmentation fault in regmatch()

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3984]
Migrated from [rt.perl.org#6987](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=6987) (status was 'resolved')

Searchable as RT6987$

**Comment by p5pRT** [comment]
### From roy@Level3.net



I was debugging some \(very\) old legacy code which has been working for
a long time\, but which failed when new data was handed to it\.

Anyway\, this code contains a somewhat insane regexp match\,
/^\.\*\(\\n\.\*\)\*$/\.  Nevermind that this expression probably isn't doing
what the coder intended \- nobody has any idea what the original intent
of this chunk of code was anymore :\-\)

This particular match works fine for small strings\, but when handed a
string containing large numbers of linefeeds\, it will cause a seg
fault\.

Here's a quick program that will trigger the problem&#8203;:

 \#\!perl
 $data = "\\n" x $ARGV\[0\];
 if \($data \!~ /^\.\*\(\\n\.\*\)\*$/\)\{ print "regexp matched\\n"; \}
 print "done\\n";

If run with a command\-line argument sufficiently large\, the problem
will occur\.  I tried perl 5\.5\.3\, 5\.6\.1 and 5\.7\.0 on Linux and Solaris
platforms\, and they all seem to be affected\.  The number of lines that
triggers the problem seems to be in the 15k\-25k range\.

A typical backtrace resembles \(this one from 5\.7\.0\)&#8203;:

\(gdb\) backtrace
\#0  0x80d3bfd in S\_regmatch \(prog=0x81054a0\) at regexec\.c&#8203;:1807
\#1  0x80d5e75 in S\_regmatch \(prog=0x81054b8\) at regexec\.c&#8203;:2770
\#2  0x80d6f9f in S\_regmatch \(prog=0x81054a0\) at regexec\.c&#8203;:3126
\#3  0x80d5e75 in S\_regmatch \(prog=0x81054b8\) at regexec\.c&#8203;:2770
\#4  0x80d6f9f in S\_regmatch \(prog=0x81054a0\) at regexec\.c&#8203;:3126
\#5  0x80d5e75 in S\_regmatch \(prog=0x81054b8\) at regexec\.c&#8203;:2770
\#6  0x80d6f9f in S\_regmatch \(prog=0x81054a0\) at regexec\.c&#8203;:3126
\#7  0x80d5e75 in S\_regmatch \(prog=0x81054b8\) at regexec\.c&#8203;:2770
\[etc\.\.\.\]

This only seems to happen with a linefeed\-based pattern\.  Substituting
another character for '\\n' in the expression will preclude triggering
of the problem\.

Thanks \!

&nbsp;                   \- roy \-

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.0:

Configured by roy at Mon Nov  6 19:40:09 UTC 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.12-20smp, archname=i686-linux
    uname='linux ug.mars.Level3.net 2.2.12-20smp #1 smp mon sep 27 10:34:45 edt 1999 i686 unknown '
    config_args='-ds -e'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2', gccversion=egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=/lib/libc-2.1.2.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/local/lib/perl5/5.6.0/i686-linux
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/roy
    LANG=en_US
    LANGUAGE (unset)
    LC_ALL=en_US
    LD_LIBRARY_PATH=/usr/local/X11R6/lib:/usr/local/X11R5/lib:/usr/local/lib:/lib:/usr/lib:/usr/openwin/lib:/export/share/snm/lib:/usr/local/lib/applix/lib
    LOGDIR (unset)
    PATH=/home/roy/util:/usr/local/bin:/usr/local/etc:/usr/bin/X11:/usr/bin/mh:/bin:/sbin:/etc:/usr/bin:/usr/sbin:/usr/etc:/home/roy/bin
    PERL_BADLANG (unset)
    SHELL=/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @vanstyn

Thanks for the report\. This is a known problem\, caused by the recursive
nature of the regexp engine\. That's something we hope to fix eventually\,
but for now\, the easiest fix would be to remove the regexp \- by its
nature\, /^\.\*\(\\n\.\*\)\*$/ must always match\.

There is a special compile\-time setting called 'REG\_INFTY' that controls
the level of recursion at which the regexp engine bails out\. This is
around 32768 by default\, but can be overridden by compile\-time switches\.
You may find it helpful to determine the linecount at which your test
program first fails on a given platform\, and recompile perl with
REG\_INFTY set slightly lower&#8203;: that will give you a chance of failing
more gracefully\. Another approach is to increase the stack size
allocated to the process\, but mechanisms for doing that vary from
system to system\.

This fragment of code from the perlre manpage will help you to determine
what REG\_INFTY is currently set to&#8203;:

&nbsp;  crypt% perl \-wle ' $\_ \*\*= $\_ \, / \{$\_\} / for 2 \.\. 42'
&nbsp;  / \{46656\} /&#8203;: Quantifier in \{\,\} bigger than 32766
&nbsp;  crypt% 

\.\. so I have REG\_INFTY = 32766 here\.

Hope this helps\,

Hugo van der Sanden

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

&nbsp;                   \- roy \-

>Reply\-to&#8203;: hv@&#8203;crypt0\.demon\.co\.uk 
>Return\-Path&#8203;: hv@&#8203;crypt\.compulink\.co\.uk \<hv@&#8203;crypt\.compulink\.co\.uk> 
>From&#8203;: Hugo \<hv@&#8203;crypt\.compulink\.co\.uk> 
>Cc&#8203;: perl5\-porters@&#8203;perl\.org 
>Date&#8203;: Mon\, 14 May 2001 15&#8203;:41&#8203;:22 \+0100 
>Subject&#8203;: \[ID 20010514\.039\] segmentation fault in regmatch\(\) 
>
>In \<200105102122\.VAA31740@&#8203;ug\.eng\.Level3\.com>\, roy@&#8203;Level3\.net writes&#8203;:
>&#8203;:
>&#8203;:This is a bug report for perl from roy@&#8203;Level3\.net\,
>&#8203;:generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
>&#8203;:
>&#8203;:\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>&#8203;:\[Please enter your report here\]
>&#8203;:
>&#8203;:Anyway\, this code contains a somewhat insane regexp match\,
>&#8203;:/^\.\*\(\\n\.\*\)\*$/\.  Nevermind that this expression probably isn't doing
>&#8203;:what the coder intended \- nobody has any idea what the original intent
>&#8203;:of this chunk of code was anymore :\-\)
>&#8203;:
>&#8203;:This particular match works fine for small strings\, but when handed a
>&#8203;:string containing large numbers of linefeeds\, it will cause a seg
>&#8203;:fault\.
>
>Thanks for the report\. This is a known problem\, caused by the recursive
>nature of the regexp engine\. That's something we hope to fix eventually\,
>but for now\, the easiest fix would be to remove the regexp \- by its
>nature\, /^\.\*\(\\n\.\*\)\*$/ must always match\.
>
>There is a special compile\-time setting called 'REG\_INFTY' that controls
>the level of recursion at which the regexp engine bails out\. This is
>around 32768 by default\, but can be overridden by compile\-time switches\.
>You may find it helpful to determine the linecount at which your test
>program first fails on a given platform\, and recompile perl with
>REG\_INFTY set slightly lower&#8203;: that will give you a chance of failing
>more gracefully\. Another approach is to increase the stack size
>allocated to the process\, but mechanisms for doing that vary from
>system to system\.
>
>This fragment of code from the perlre manpage will help you to determine
>what REG\_INFTY is currently set to&#8203;:
>
>  crypt% perl \-wle ' $\_ \*\*= $\_ \, / \{$\_\} / for 2 \.\. 42'
>  / \{46656\} /&#8203;: Quantifier in \{\,\} bigger than 32766
>  crypt% 
>
>\.\. so I have REG\_INFTY = 32766 here\.
>
>Hope this helps\,
>
>Hugo van der Sanden

**Comment by p5pRT** [comment]
### From @smpeters

> \[roy@&#8203;Level3\.net \- Sun May 13 22&#8203;:41&#8203;:36 2001\]&#8203;:
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> I was debugging some \(very\) old legacy code which has been working for
> a long time\, but which failed when new data was handed to it\.
> 
> Anyway\, this code contains a somewhat insane regexp match\,
> /^\.\*\(\\n\.\*\)\*$/\.  Nevermind that this expression probably isn't doing
> what the coder intended \- nobody has any idea what the original intent
> of this chunk of code was anymore :\-\)
> 
> This particular match works fine for small strings\, but when handed a
> string containing large numbers of linefeeds\, it will cause a seg
> fault\.
> 
> Here's a quick program that will trigger the problem&#8203;:
> 
>  \#\!perl
>  $data = "\\n" x $ARGV\[0\];
>  if \($data \!~ /^\.\*\(\\n\.\*\)\*$/\)\{ print "regexp matched\\n"; \}
>  print "done\\n";
> 
> If run with a command\-line argument sufficiently large\, the problem
> will occur\.  I tried perl 5\.5\.3\, 5\.6\.1 and 5\.7\.0 on Linux and Solaris
> platforms\, and they all seem to be affected\.  The number of lines that
> triggers the problem seems to be in the 15k\-25k range\.
> 
> A typical backtrace resembles \(this one from 5\.7\.0\)&#8203;:
> 
> \(gdb\) backtrace
> \#0  0x80d3bfd in S\_regmatch \(prog=0x81054a0\) at regexec\.c&#8203;:1807
> \#1  0x80d5e75 in S\_regmatch \(prog=0x81054b8\) at regexec\.c&#8203;:2770
> \#2  0x80d6f9f in S\_regmatch \(prog=0x81054a0\) at regexec\.c&#8203;:3126
> \#3  0x80d5e75 in S\_regmatch \(prog=0x81054b8\) at regexec\.c&#8203;:2770
> \#4  0x80d6f9f in S\_regmatch \(prog=0x81054a0\) at regexec\.c&#8203;:3126
> \#5  0x80d5e75 in S\_regmatch \(prog=0x81054b8\) at regexec\.c&#8203;:2770
> \#6  0x80d6f9f in S\_regmatch \(prog=0x81054a0\) at regexec\.c&#8203;:3126
> \#7  0x80d5e75 in S\_regmatch \(prog=0x81054b8\) at regexec\.c&#8203;:2770
> \[etc\.\.\.\]
> 
> This only seems to happen with a linefeed\-based pattern\.  Substituting
> another character for '\\n' in the expression will preclude triggering
> of the problem\.
> 
> Thanks \!

This has been fixed with change \#27598\.\.\.

steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ ulimit \-s 1024
steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ perl rt\_6987\.pl 4096
Segmentation fault
steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ \./perl rt\_6987\.pl 4096
done
steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ \./perl rt\_6987\.pl 16384
done

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
